nixos/modules/services/web-servers/tomcat/default.nix

```diff
@@ -0,0 +1,554 @@
+      sslServerKey = if useACME then "${sslCertDir}/key.pem" else hostOpts.sslServerKey;
+      sslServerChain = if useACME then "${sslCertDir}/fullchain.pem" else hostOpts.sslServerChain;
+
+      # https://ssl-config.mozilla.org/#server=tomcat&version=9.0.30&config=intermediate&guideline=5.4
+      defaultAttrs = {
+        inherit (hostOpts) hostName;
+
+        ciphers = "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384";
+        disableSessionTickets = true;
+        honorCipherOrder = false;
+        protocols = concatStringsSep " " ([ "TLSv1.2" ] ++ optional (versionAtLeast cfg.jdk.version "11.0") "TLSv1.3");
+      };
+    in
+      ''
+        <SSLHostConfig ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") (defaultAttrs // hostOpts.sslHostConfig))}>
+          <Certificate
+            certificateKeyFile="${sslServerKey}"
+            certificateFile="${sslServerCert}"
+            certificateChainFile="${sslServerChain}"
+          />
+        </SSLHostConfig>
+      ''
+  ;
+
+  mkVHostConf = hostOpts:
+    let
+      useACME = hostOpts.enableACME; # TODO: don't try to serve up directories that don't exist - ie. are we using DNS challenge? || hostOpts.useACMEHost != null;
+      acmeChallenge = optionalString useACME ''
+        <Context docBase="${hostOpts.acmeRoot}/.well-known/acme-challenge" path="/.well-known/acme-challenge" />
+      '';
+      robotsEntries = optionalString (hostOpts.robotsEntries != "") ''
+        <Context docBase="${pkgs.writeText "robots.txt" hostOpts.robotsEntries}" path="/robots.txt" />
+      '';
+
+      # the following attributes allow tomcat to function with an immutable configuration directory
+      # see: https://tomcat.apache.org/tomcat-9.0-doc/config/host.html
+      defaultAttrs = {
+        name = hostOpts.hostName;
+        appBase = hostOpts.appBase;
+        unpackWARs = true;
+        autoDeploy = true;
+        xmlBase = "${stateDir}/CATALINA/${hostOpts.hostName}";
+        workDir = "${cacheDir}/${hostOpts.hostName}";
+      };
+    in
+      ''
+        <Host ${concatStringsSep " " (mapAttrsToList (k: v: "${k}=\"${toStr v}\"") (defaultAttrs // hostOpts.settings))}>
+          ${concatMapStrings (alias: "<Alias>${alias}</Alias>") hostOpts.serverAliases}
+          ${acmeChallenge}
+          ${robotsEntries}
+          ${optionalString (hostOpts.logFormat != null) ''
+            <Valve
+              className="org.apache.catalina.valves.AccessLogValve"
+              directory="${cfg.logDir}"
+              prefix="${hostOpts.hostName}_access" suffix=".log"
+              pattern="${hostOpts.logFormat}"
+            />
+          ''}
+          ${hostOpts.extraConfig}
+        </Host>
+      ''
+  ;
+
+  serverXmlFile = pkgs.writeText "server.xml" ''
```

Comment:
by the way `toXML` does not work for cases like this?

Reply:
I'll have to give that a try. Thanks @Mic92!